### PR TITLE
Maintenance controller

### PR DESCRIPTION
**Pull request by adrianchiris**
Refer to individual commits for context.

### CODE REVIEW

File: cmd/maintenance-manager/main.go [hunk 1]
```diff
@@ -132,20 +138,30 @@ func main() {
 		os.Exit(1)
 	}
 
+	k8sInterface, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		setupLog.Error(err, "unable to create kubernetes interface")
+		os.Exit(1)
+	}
+
+	mgrClient := mgr.GetClient()
```
Comment (ykulazhenkov) [inline cmd/maintenance-manager/main.go:147]: Do we need to explicitly share the client? I think mgr.GetClient() will always return the same instance from the manager

File: internal/controller/nodemaintenance_controller.go [hunk 2]
```diff
@@ -98,64 +118,342 @@ type NodeMaintenanceReconciler struct {
...
+	return ReadyConditionChangedPredicate{
+		Funcs: predicate.Funcs{},
+		log:   log,
+	}
+}
+
+// ReadyConditionChangedPredicate will trigger enqueue of Event for reconcile in the following cases:
+// 1. A change in NodeMaintenance Ready Condition
+// 2. Update to the object occurred and deletion timestamp is set
+// 3 NodeMaintenance created
```
Comment (ykulazhenkov) [inline internal/controller/nodemaintenance_controller.go:416]: nit:  add dot after `3`

File: internal/controller/nodemaintenance_controller_test.go [hunk 3]
```diff
@@ -116,24 +156,82 @@ var _ = Describe("NodeMaintenance Controller", func() {
...
+
+			By("Eventually NodeMaintenance condition is set to Pending")
+			Eventually(testutils.GetReadyConditionReasonForFn(testCtx, k8sClient, client.ObjectKeyFromObject(nm))).
+				WithTimeout(10 * time.Second).WithPolling(1 * time.Second).Should(Equal(maintenancev1.ConditionReasonPending))
+
+			By("Set NodeMaintenance to Scheduled")
+			Expect(k8sClient.Get(testCtx, types.NamespacedName{Namespace: "default", Name: "test-nm"}, nm)).ToNot(HaveOccurred())
+			Expect(k8sutils.SetReadyConditionReason(testCtx, k8sClient, nm, maintenancev1.ConditionReasonScheduled)).ToNot(HaveOccurred())
+
+			By("Eventually NoddeMaintenance condition is set to WaitForPodCompletion")
```
Comment (ykulazhenkov) [inline internal/controller/nodemaintenance_controller_test.go:187]: NoddeMaintenance > NodeMaintenance

**Review by ykulazhenkov — COMMENTED** [review]
good job! Added few minor comments

File: cmd/maintenance-manager/main.go (hunk 1 above)
Comment (adrianchiris) [reply cmd/maintenance-manager/main.go:147]: we dont, its just shorter to specify than mgr.GetClient() i can change if thats your preference LMK

File: internal/podcompletion/podcompletion.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	maintenancev1 "github.com/Mellanox/maintenance-operator/api/v1alpha1"
+)
+
+const (
+	WaitForPodCompletionStartAnnot string = "maintenance.nvidia.com/wait-pod-completion-start-time"
```
[the remark this answers is not in the document]
Comment (adrianchiris) [reply internal/podcompletion/podcompletion.go:36]: we assume time is synchronized across nodes, at least control plane where we would like this to run.

lastTransitionTime is also set by the controller, part of metav1 packakge in SetStatusCondition.
see[1]

[1]https://github.com/kubernetes/apimachinery/blob/95b78024e3feada7739b40426690b4f287933fd8/pkg/api/meta/conditions.go#L38

File: internal/controller/nodemaintenance_controller.go (hunk 2 above)
Comment (adrianchiris) [reply internal/controller/nodemaintenance_controller.go:416]: done.

File: internal/controller/nodemaintenance_controller_test.go (hunk 3 above)
Comment (adrianchiris) [reply internal/controller/nodemaintenance_controller_test.go:187]: bad keyboard key doing double typing, fixed.

File: internal/podcompletion/podcompletion_suite_test.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+package podcompletion_test
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestUtils(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (adrianchiris) [reply internal/podcompletion/podcompletion_suite_test.go:26]: done.

File: internal/cordon/cordon_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+package cordon_test
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestUtils(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (adrianchiris) [reply internal/cordon/cordon_suite_test.go:26]: done.

File: cmd/maintenance-manager/main.go (hunk 1 above)
Comment (ykulazhenkov) [reply cmd/maintenance-manager/main.go:147]: No It is ok to keep it like this, I just want to check that I'm not missing anything.

File: internal/podcompletion/podcompletion.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ykulazhenkov) [reply internal/podcompletion/podcompletion.go:36]: Thanks, for the clarification.

**Review by ykulazhenkov — APPROVED** [review]
LGTM

**Pull request merged**